File: Data Structures & Algorithms/palindromic-substrings/submission-1.cpp
class Solution {
public:
    int countSubstrings(string s) {
        int ans = 0;
        int size = s.length();
        vector<vector<bool>> dp(size, vector<bool>(size, false));
        
        for (int i = size - 1; i >= 0; i--) {
            for (int j = i; j < size; j++) {
                if (s[i] == s[j] && (j - i <= 2 || dp[i + 1][j - 1])) {
                    ans += 1;
                    dp[i][j] = true;
                }
            }
        }
        return ans;
    }
};

File: Data Structures & Algorithms/count-subsequences/submission-1.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        memo = {}

        def dfs(i, j):
            if j == len(t):
                return 1
            if i == len(s):
                return 0
            if (i, j) in memo:
                return memo[(i, j)]
            ans = 0
            if s[i] == t[j]:
                ans += dfs(i + 1, j + 1)
            ans += dfs(i + 1, j)

            memo[(i, j)] = ans
            return ans
        
        return dfs(0, 0)